src/vsr/journal.zig

```diff
@@ -2142,25 +2145,34 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
             return sector;
         }
 
-        pub fn writing(journal: *Journal, op: u64, checksum: u128) bool {
-            const slot = journal.slot_for_op(op);
-            var found: bool = false;
-            var it = journal.writes.iterate();
-            while (it.next()) |write| {
-                const write_slot = journal.slot_for_op(write.message.header.op);
+        const Writing = enum {
+            none,
+            /// Either the prepare or the redundant header of a message with the same slot as the
+            /// given op is being written. It may be a different version of the same op, or a
+            /// different op which shares the prepare slot.
+            slot,
+            /// Either the prepare or the redundant header of a message with the exact op/checksum
+            /// is being written.
+            exact,
+        };
 
-                // It's possible that we might be writing the same op but with a different checksum.
-                // For example, if the op we are writing did not survive the view change and was
-                // replaced by another op. We must therefore do the search primarily on checksum.
-                // However, we compare against the 64-bit op first, since it's a cheap machine word.
-                if (write.message.header.op == op and write.message.header.checksum == checksum) {
-                    // If we truly are writing, then the dirty bit must be set:
-                    assert(journal.dirty.bit(journal.slot_for_op(op)));
-                    found = true;
-                } else if (write_slot.index == slot.index) {
-                    // If the in-progress write of '{op, checksum}' will be overwritten by another
-                    // write to the same slot, writing() must return false.
-                    found = false;
+        pub fn writing(journal: *Journal, header: *const Header.Prepare) Writing {
+            const slot = journal.slot_for_header(header);
+            var found: Writing = .none;
+            var writes = journal.writes.iterate();
+            while (writes.next()) |write| {
+                const write_slot = journal.slot_for_op(write.message.header.op);
+                if (write.message.header.op == header.op and
+                    write.message.header.checksum == header.checksum)
```

Comment:
Might want to keep the comment explaining why we compare both op & checksum

Reply:
Though, I am not 100% convinced that the reasoning is sound. We might as well branch after comparing the first 16 bytes of checksum! So another option is to compare just the sums, and assert op equality.